src/chain.h

```diff
@@ -332,31 +330,25 @@ class CDiskBlockIndex : public CBlockIndex
         hashPrev = (pprev ? pprev->GetBlockHash() : uint256());
     }
 
-    ADD_SERIALIZE_METHODS;
-
-    template <typename Stream, typename Operation>
-    inline void SerializationOp(Stream& s, Operation ser_action) {
+    SERIALIZE_METHODS(CDiskBlockIndex, obj)
+    {
         int _nVersion = s.GetVersion();
-        if (!(s.GetType() & SER_GETHASH))
-            READWRITE(VARINT(_nVersion, VarIntMode::NONNEGATIVE_SIGNED));
-
-        READWRITE(VARINT(nHeight, VarIntMode::NONNEGATIVE_SIGNED));
-        READWRITE(VARINT(nStatus));
-        READWRITE(VARINT(nTx));
-        if (nStatus & (BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO))
-            READWRITE(VARINT(nFile, VarIntMode::NONNEGATIVE_SIGNED));
-        if (nStatus & BLOCK_HAVE_DATA)
-            READWRITE(VARINT(nDataPos));
-        if (nStatus & BLOCK_HAVE_UNDO)
-            READWRITE(VARINT(nUndoPos));
+        if (!(s.GetType() & SER_GETHASH)) READWRITE(VARINT(_nVersion, VarIntMode::NONNEGATIVE_SIGNED));
+
+        READWRITE(VARINT(obj.nHeight, VarIntMode::NONNEGATIVE_SIGNED));
+        READWRITE(VARINT(obj.nStatus));
+        READWRITE(VARINT(obj.nTx));
+        if (obj.nStatus & (BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO)) READWRITE(VARINT(obj.nFile, VarIntMode::NONNEGATIVE_SIGNED));
+        if (obj.nStatus & BLOCK_HAVE_DATA) READWRITE(VARINT(obj.nDataPos));
+        if (obj.nStatus & BLOCK_HAVE_UNDO) READWRITE(VARINT(obj.nUndoPos));
```

Comment:
9b66083788581c264a097e26795561cb3eac455d

I know people aren't fans of macros so this will probably be shut down, but this pattern appears often enough to almost warrant a `READWRITEIF(condition, ...)`. Above cases would be
```C++
READWRITEIF(!(s.GetType() & SER_GETHASH), VARINT(_nVersion, VarIntMode::NONNEGATIVE_SIGNED));
...
READWRITEIF(obj.nStatus & (BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO), VARINT(obj.nFile, VarIntMode::NONNEGATIVE_SIGNED));
READWRITEIF(obj.nStatus & BLOCK_HAVE_DATA, VARINT(obj.nDataPos));
READWRITEIF(obj.nStatus & BLOCK_HAVE_UNDO, VARINT(obj.nUndoPos));
```

Reply:
In general I'm not opposed to adding more macros for commonly-used serialization patterns, but I think that the benefit of "READWRITEIF(a, b)" over "if (a) READWRITE(b)" is marginal at best.